File: frontend/src/marvel/types/EncounterCard.ts
import { JsonDecoder } from 'ts.data.json'
import { CardDef, cardDefDecoder } from '@/marvel/types/CardDef'

export interface EncounterCard {
  cardId: string
  cardDef: CardDef
}

export const encounterCardDecoder = JsonDecoder.object<EncounterCard>({
  cardId: JsonDecoder.string,
  cardDef: cardDefDecoder
}, 'EncounterCard', { cardId: 'ecCardId', cardDef: 'ecCardDef' })

File: backend/marvel-core/library/Marvel/Card/EncounterCard.hs
module Marvel.Card.EncounterCard where

import Marvel.Prelude

import Marvel.Card.Code
import Marvel.Card.Def
import Marvel.Card.Id

data EncounterCard = MkEncounterCard
  { ecCardId :: CardId
  , ecCardDef :: CardDef
  }
  deriving stock (Show, Eq, Generic)
  deriving anyclass (ToJSON, FromJSON, Hashable)

instance HasCardId EncounterCard where
  toCardId = ecCardId

instance HasCardCode EncounterCard where
  toCardCode = toCardCode . getCardDef

instance HasCardDef EncounterCard where
  getCardDef = ecCardDef
